### PR TITLE
Add FANOUT routing policy

### PR DESCRIPTION
**Pull request by gierlachg**
Added `FANOUT` routing policy.
For distributed edge case it sends each item to a single processor on each of the cluster members - a combination of `BROADCAST` & `ROUND_ROBIN`. For a local edge case it behaves exactly as `ROUND_ROBIN`.

Useful for optimized `IMap` (SQL) join implementations where single item is being joined with local `IMap` partitions.

Checklist:
- [x] Labels and Milestone set
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java [hunk 1]
```diff
@@ -676,7 +692,12 @@ public int getClassId() {
         /**
          * This policy sends each item to all candidate processors.
          */
-        BROADCAST
+        BROADCAST,
+        /**
+         * This policy sends each item to a single processor on each of the
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java:697]: This doesn't state which processor on each member. The PR mentions "round robin" so I guess it's a different processor every time?

Comment (gierlachg) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java:697]: Slightly reworded.

Comment (ufukyilmaz) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java:697]: I think it would be better to describe the procedure made in a round-robin fashion in another sentence. Because it's a bit unclear what exactly is done in a round-robin fashion.

Comment (mtopolnik) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java:697]: Let's follow the lead of UNICAST, which says "with no restriction on the choice." Here we can say "sends each item to a single processor on every cluster member. Which exactly processor inside a member it chooses is arbitrary and may change with every item.".

Comment (ufukyilmaz) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java:697]: Currently, we can only run a DAG containing a fanout edge in a cluster containing multiple members. Is this expected behavior? I would expect it to act like unicast when it is running on single member because we can normally run a DAG with a distributed edge on a single member.
Otherwise, failing the job in this case is the expected behavior, let's note it here. Also, I think that we should add tests to RoutingPolicyTest for distributed edge types to demonstrate the single member behavior of them (I am aware that currently other distributed edge types isn't tested for their single member behaviors either.)

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java [hunk 2]
```diff
@@ -378,23 +379,160 @@ private void initDag(InternalSerializationService jobSerializationService) {
...
+            EdgeDef edge,
+            int processorIndex,
+            InternalSerializationService jobSerializationService
+    ) {
+        if (edge.isDistributed() && edge.routingPolicy() == RoutingPolicy.ISOLATED) {
+            throw new IllegalArgumentException("Isolated edges must be local: " + edge);
+        }
+        if ((!edge.isDistributed() || !edge.getDistributedTo().equals(DISTRIBUTE_TO_ALL))
+            && edge.routingPolicy() == RoutingPolicy.FANOUT) {
+            throw new IllegalArgumentException("Fanout edges must be distributed: " + edge);
```
Comment (ufukyilmaz) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java:414]: This causes a job to be failed when it runs on a single member. If it is not intended, you may check [this](https://github.com/hazelcast/hazelcast-jet/blob/master/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/EdgeDef.java#L58)

**Comment by ufukyilmaz** [comment]
I think it would be better to do the code relocations in a different PRs since these make it difficult to review.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java [hunk 3]
```diff
@@ -378,23 +379,160 @@ private void initDag(InternalSerializationService jobSerializationService) {
...
-                        memberToSenderConveyorMap(edgeSenderConveyorMap, edge, jobSerializationService);
+    private OutboundCollector createOutboundCollector(
+            EdgeDef edge,
+            int processorIndex,
+            InternalSerializationService jobSerializationService
+    ) {
+        if (edge.isDistributed() && edge.routingPolicy() == RoutingPolicy.ISOLATED) {
+            throw new IllegalArgumentException("Isolated edges must be local: " + edge);
+        }
+        if ((!edge.isDistributed() || !edge.getDistributedTo().equals(DISTRIBUTE_TO_ALL))
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java:412]: If the second condition is true, the first is always true:
```suggestion
        if (!DISTRIBUTE_TO_ALL.equals(edge.getDistributedTo())
```

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/EdgeDef.java [hunk 4]
```diff
@@ -114,6 +114,10 @@ Address getDistributedTo() {
         return distributedTo;
     }
 
+    boolean isDistributed() {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/EdgeDef.java:117]: I specifically didn't want to create this method because of the distributed-to-one edge type - it won't strike your eye that there are two types of distributed edges, especially when the distributed-to-one is rarely used. Maybe we can create `isLocal` method instead, it seems that there are more instances of `!edge.isDistributed()` than without the negation.

**Review by viliam-durina — APPROVED** [review]
My comments are regarding the readability, so I'm approving.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java [hunk 5]
```diff
@@ -676,7 +692,13 @@ public int getClassId() {
...
+         * cluster members in a round robin fashion. It is only available
+         * on a distributed edge.
+         */
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java:700]:
```suggestion
        /**
         * This policy sends an item to all members, but only to one processor on
         * each member. It's a combination of {@link #BROADCAST} and {@link
         * #UNICAST}: an item is first <em>broadcast</em> to all members, and then,
         * on each member, it is <em>unicast</em> to one processor.
         * <p>
         * If the destination local parallelism is 1, the behavior is equal to
         * {@link #BROADCAST}.
         *
         * @since 4.4
         */
```

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3842/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3842/jet-oss-pr-builder-3842.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   StreamKafkaPTest.when_partitionAddedWhileJobDown_then_consumedFromBeginning:403->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248->lambda$when_partitionAddedWhileJobDown_then_consumedFromBeginning$8:404 expected:<1> but was:<0>
[INFO] 
[ERROR] Tests run: 35, Failures: 1, Errors: 0, Skipped: 0
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3843/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3843/jet-oss-pr-builder-3843.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   StreamKafkaPTest.when_partitionAddedWhileJobDown_then_consumedFromBeginning:403->HazelcastTestSupport.assertTrueEventually:1350->HazelcastTestSupport.assertTrueEventually:1248->lambda$when_partitionAddedWhileJobDown_then_consumedFromBeginning$8:404 expected:<1> but was:<2>
[INFO] 
[ERROR] Tests run: 35, Failures: 1, Errors: 0, Skipped: 0
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Review by ufukyilmaz — APPROVED** [review]

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java [hunk 6]
```diff
@@ -676,7 +690,19 @@ public int getClassId() {
...
-        BROADCAST
+        BROADCAST,
+        /**
+         * This policy sends an item to all members, but only to one processor on
+         * each member. It's a combination of {@link #BROADCAST} and {@link
+         * #UNICAST}: an item is first <em>broadcast</em> to all members, and then,
+         * on each member, it is <em>unicast</em> to one processor.
+         * <p>
+         * If the destination local parallelism is 1, the behavior is equal to
+         * {@link #BROADCAST}.
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/Edge.java:701]:
```suggestion
         * If the destination local parallelism is 1, the behavior is equal to
         * {@link #BROADCAST}. If the member count in the cluster is 1, the
         * behavior is equal to {@link #UNICAST}.
```

**Review by viliam-durina — APPROVED** [review]
